### PR TITLE
316 disable button 0 qtum

### PR DESCRIPTION
**Pull request by alex-cory**
Added
- `not enough BOT or QTUM` error
- instant error message + disable button for `create event`, `bet`, `vote`, `result setting`
- refactors

### CODE REVIEW

File: src/components/EventWarning/index.js [hunk 1]
```diff
@@ -4,47 +4,31 @@ import { withStyles } from 'material-ui/styles';
...
-      message,
-      intl,
-      theme,
-      ...props
-    } = this.props;
 
-    if (!message) {
-      return null;
-    }
+const FormattedMessageFixed = (props) => <FormattedMessage {...props} />; // workaround: http://bit.ly/2u7Hhe4
```
Comment (liviavinci) [inline src/components/EventWarning/index.js]: why is this necessary?

File: src/scenes/CreateEvent/components/CreateEventCreatorPicker/index.js [hunk 2]
```diff
@@ -26,10 +26,14 @@ class CreateEventCreatorPicker extends React.PureComponent {
     lastUsedAddress: PropTypes.string.isRequired,
     setLastUsedAddress: PropTypes.func.isRequired,
     changeFormFieldValue: PropTypes.func.isRequired,
-    eventEscrowAmount: PropTypes.number.isRequired,
+    eventEscrowAmount: PropTypes.number,
     intl: intlShape.isRequired, // eslint-disable-line react/no-typos
   };
 
+  static defaultProps = {
+    eventEscrowAmount: 100,
```
Comment (liviavinci) [inline src/scenes/CreateEvent/components/CreateEventCreatorPicker/index.js]: let's revert back to the required since 100 is not accurate as the default value

File: src/scenes/CreateEvent/index.js [hunk 3]
```diff
@@ -282,88 +356,53 @@ class CreateEvent extends React.Component {
             </div>
             <Grid container>
               <Grid item xs={3}>
-                {intl.formatMessage(messages.creator)}
+                <DialogTitle className={classes.title}>
```
Comment (liviavinci) [inline src/scenes/CreateEvent/index.js]: The relationship between dialog title and body should be siblings

File: src/scenes/CreateEvent/index.js [hunk 4]
```diff
@@ -129,35 +131,125 @@ const messages = defineMessages({
...
     getInsightTotals: undefined,
     eventEscrowAmount: undefined,
-  };
+  }
 
   state = {
     selectAddressDialogVisibility: false,
-  };
+    error: {},
+    isValid: true,
```
Comment (liviavinci) [inline src/scenes/CreateEvent/index.js]: needs a more descriptive name specifying that it's for the non zero qtum check, like hasEnoughQTUM

File: src/scenes/CreateEvent/index.js [hunk 5]
```diff
@@ -384,99 +423,29 @@ class CreateEvent extends React.Component {
...
-
-// decorate with redux-form
-const createEventForm = reduxForm({
-  form: FORM_NAME,
-  fields,
-  validate,
-})(CreateEvent);
-
-export default connect(mapStateToProps, mapDispatchToProps)(injectIntl(withStyles(styles, { withTheme: true })(createEventForm)));
+const Section = injectIntl(({ title, children, intl }) => (
```
Comment (liviavinci) [inline src/scenes/CreateEvent/index.js]: awesome!

File: src/scenes/Event/scenes/topic.js [hunk 6]
```diff
@@ -409,10 +406,8 @@ export default class TopicPage extends React.Component {
       return {
         show: true,
         disabled: true,
-        message: <FormattedMessage
-          id="str.pendingTransactionDisabledMsg"
-          defaultMessage="You have a pending transaction for this event. Please wait until it's confirmed before doing another transaction."
-        />,
+        id: 'str.pendingTransactionDisabledMsg',
```
Comment (liviavinci) [inline src/scenes/Event/scenes/topic.js]: the return format is changed but not the place where it's used?

File: src/scenes/Event/scenes/oracle.js [hunk 7]
```diff
@@ -589,74 +538,77 @@ export default class OraclePage extends React.Component {
...
     }
 
     // Not enough bot for setting the result or voting
-    const filteredAddress = _.filter(walletAddresses, { address: lastUsedAddress });
-    const currentBot = filteredAddress.length > 0 ? filteredAddress[0].bot : 0;
     if ((token === Token.Qtum
       && (status === OracleStatus.WaitResult || status === OracleStatus.OpenResultSet)
-      && currentBot < oracle.consensusThreshold)
-      || (token === Token.Bot && status === OracleStatus.Voting && currentBot < voteAmount)) {
+      && totalBot < oracle.consensusThreshold)
```
Comment (liviavinci) [inline src/scenes/Event/scenes/oracle.js]: Revert. BOT & QTUM works differently. You can spend qtum from different addresses for the same tx but not BOT. That's why you are not supposed to compare with totalBot.

File: src/components/EventWarning/index.js (hunk 1 above)
Comment (alex-cory) [reply src/components/EventWarning/index.js]: did you reqad the url in the comments? :)

Comment (liviavinci) [reply src/components/EventWarning/index.js]: somehow the url doesn't work for me

File: src/scenes/CreateEvent/index.js [hunk 8]
```diff
@@ -193,66 +285,51 @@ class CreateEvent extends React.Component {
...
-  };
+  }
 
   onClose = () => {
     this.props.toggleCreateEventDialog(false);
-  };
+  }
+
+  validateEnoughQTUM = () => {
+    const totalQtum = _.sumBy(this.props.walletAddresses, ({ qtum }) => qtum ? qtum : 0); // eslint-disable-line
```
Comment (dwalintukan) [inline src/scenes/CreateEvent/index.js]: can we do this without disabling the eslint?

File: src/scenes/Event/scenes/oracle.js [hunk 9]
```diff
@@ -589,74 +539,77 @@ export default class OraclePage extends React.Component {
...
-    }
-
-    // Did not enter an amount
-    if (status === OracleStatus.Voting && (voteAmount <= 0 || Number.isNaN(voteAmount))) {
+    // Trying to set result or vote when not enough QTUM or BOT
+    const totalQtum = _.sumBy(walletAddresses, (wallet) => wallet.qtum ? wallet.qtum : 0);
+    const filteredAddress = _.filter(walletAddresses, { address: lastUsedAddress });
+    const currentBot = filteredAddress.length > 0 ? filteredAddress[0].bot : 0; // # of BOT at currently selected address
+    if ((token === Token.Bot || (token === Token.Qtum && status === OracleStatus.OpenResultSet))
+      && (totalQtum < maxTransactionFee && currentBot < oracle.consensusThreshold)
```
Comment (dwalintukan) [inline src/scenes/Event/scenes/oracle.js]: `currentBot < oracle.consensusThreshold` is only applicable when the user is setting the result. `token === Token.Bot` + `status === OracleStatus.Voting` (which is Voting status) should check `currentBot < voteAmount`.

Rules for allowing action by status:
- Bet: `totalQtum >= voteAmount + maxTransactionFee`
- SetResult: `totalQtum >= maxTransactionFee` + `currentBot >= consensusThreshold`
- Vote: `totalQtum >= maxTransactionFee` + `currentBot >= voteAmount`
- Finalize: `totalQtum >= maxTransactionFee`
- Withdraw: `totalQtum >= maxTransactionFee`

File: src/scenes/Event/scenes/oracle.js [hunk 10]
```diff
@@ -589,74 +539,77 @@ export default class OraclePage extends React.Component {
...
+        id: 'str.notEnoughQtumAndBot',
+        message: 'You don\'t have enough QTUM or BOT',
+        warningTypeClass: EventWarningType.Error,
       };
     }
 
     // Trying to bet more qtum than you have
-    const totalQtum = _.sumBy(walletAddresses, (wallet) => wallet.qtum ? wallet.qtum : 0);
-    if (token === Token.Qtum && status === OracleStatus.Voting && voteAmount > totalQtum) {
+    if ((status === OracleStatus.Voting && voteAmount > totalQtum + maxTransactionFee)
```
Comment (dwalintukan) [inline src/scenes/Event/scenes/oracle.js]: you need to check the token type.
`token === Token.Qtum && status === OracleStatus.Voting` = Bet phase
`token === Token.Bot && status === OracleStatus.Voting` = Vote phase

see above comment about rules

File: src/scenes/Event/scenes/oracle.js [hunk 11]
```diff
@@ -589,74 +539,77 @@ export default class OraclePage extends React.Component {
...
     }
 
     // Not enough bot for setting the result or voting
-    const filteredAddress = _.filter(walletAddresses, { address: lastUsedAddress });
-    const currentBot = filteredAddress.length > 0 ? filteredAddress[0].bot : 0;
     if ((token === Token.Qtum
       && (status === OracleStatus.WaitResult || status === OracleStatus.OpenResultSet)
       && currentBot < oracle.consensusThreshold)
-      || (token === Token.Bot && status === OracleStatus.Voting && currentBot < voteAmount)) {
+      || (token === Token.Bot && status === OracleStatus.Voting && (currentBot < voteAmount || currentBot < oracle.consensusThreshold))) {
```
Comment (dwalintukan) [inline src/scenes/Event/scenes/oracle.js]: `currentBot < oracle.consensusThreshold` is invalid here. If the user wants to vote 1 BOT, why does it matter if their `currentBot` is under the threshold?

File: .eslintrc.json [hunk 12]
```diff
@@ -22,6 +22,8 @@
   },
   "rules": {
     "object-curly-newline": 0,
+    "react/no-did-mount-set-state": 0,
+    "react/jsx-pascal-case": 0,
```
[the remark this answers is not in the document]
Comment (alex-cory) [reply .eslintrc.json]: https://reactjs.org/docs/react-component.html#componentdidmount
image

[the remark this answers is not in the document]
Comment (alex-cory) [reply .eslintrc.json]: https://github.com/airbnb/javascript/issues/684#issuecomment-172379370

File: src/scenes/CreateEvent/index.js (hunk 8 above)
Comment (alex-cory) [reply src/scenes/CreateEvent/index.js]: yep! Just tested in dev console on lodash site.
```js
var list = [
    {
		qtum: 1
    },
    {
		qtum: 3
    },
]
var totalQTUM = _.sumBy(list, ({ qtum }) => qtum)
console.log('TOTAL: ', totalQTUM)
TOTAL:  4
```

File: .eslintrc.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (dwalintukan) [reply .eslintrc.json]: good to know thanks 👍

File: src/scenes/CreateEvent/index.js (hunk 8 above)
Comment (dwalintukan) [reply src/scenes/CreateEvent/index.js]: cool. I just feel like we should obey the preset ESLint rules unless we have a good reason to disable a rule, which is not just for convenience's sake.

File: src/scenes/CreateEvent/index.js [hunk 13]
```diff
@@ -1,28 +1,30 @@
...
 import _ from 'lodash';
 import moment from 'moment';
 import Web3Utils from 'web3-utils';
-import { Field, FieldArray, reduxForm, Form, formValueSelector, change, reset } from 'redux-form';
+import { Field, reduxForm, Form, formValueSelector, change, reset } from 'redux-form';
 import Dialog, { DialogActions, DialogContent, DialogTitle } from 'material-ui/Dialog';
 import Grid from 'material-ui/Grid';
-import TextField from 'material-ui/TextField';
-import Input, { InputLabel, InputAdornment } from 'material-ui/Input';
+import _TextField from 'material-ui/TextField';
```
[the remark this answers is not in the document]
Comment (alex-cory) [reply src/scenes/CreateEvent/index.js]: Yeah! If you look at line 445, it's just because then we can use the same name. We're making our own stateless component to essentially accomplish the same goal.

File: src/scenes/CreateEvent/index.js [hunk 14]
```diff
@@ -193,66 +285,52 @@ class CreateEvent extends React.Component {
...
-  };
+  }
 
   onClose = () => {
     this.props.toggleCreateEventDialog(false);
-  };
+  }
+
+  validateEnoughQTUM = () => {
+    // debugger; // eslint-disable-line
```
[the remark this answers is not in the document]
Comment (alex-cory) [reply src/scenes/CreateEvent/index.js]: oops sorry lol
